// Copyright 2018, Intel Corp.

syntax = "proto3";

package vertexai.tile.stripe.proto;

import "tile/proto/shape.proto";

message Block {
  string name = 1;
  string comments = 2;
  repeated string index_names = 3;
  repeated uint64 index_ranges = 4;
  repeated Constraint constraints = 5;
  repeated Declaration decls = 6;
  repeated RefineIn ref_ins = 7;
  repeated RefineOut ref_outs = 8;
  repeated Statement stmts = 9;
}

message Declaration {
  string name = 1;
  shape.proto.TensorShape shape = 2;
}

message RefineIn {
  string name = 1;
  BufferAccess access = 2;
}

message RefineOut {
  string name = 1;
  BufferAccess access = 2;
  string agg_op = 3;
}

message BufferAccess {
  int64 offset = 1;
  repeated int64 strides = 2;
}

message Statement {
  oneof op {
    Load load = 1;
    Store store = 2;
    Intrinsic intrinsic = 3;
    Constant constant = 4;
    Block block = 5;
    Special special = 6;
  }
}

message Load {
  string from = 1;
  string into = 2;
}

message Store {
  string from = 1;
  string into = 2;
}

message Intrinsic {
  string name = 1;
  repeated string inputs = 2;
  repeated string outputs = 3;
}

message Special {
  string name = 1;
  repeated string params = 2;
  repeated string inputs = 3;
  repeated string outputs = 4;
}

message Constant {
  string name = 1;
  oneof value {
    int64 iconst = 2;
    double fconst = 3;
  }
}

message Constraint {
  repeated int64 lhs = 1;
  int64 rhs = 2;
}
